pkg/snmp/traps/server.go

```diff
@@ -0,0 +1,170 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2020 Datadog, Inc.
+
+package traps
+
+import (
+	"net"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/soniah/gosnmp"
+)
+
+/*
+* Traps pipeline diagram
+* ----------------------
+*
+* (devices) --[udp:port1]-!-> (Listener 1) \
+*                         !                 (TrapServer) --[chan]--> (Forwarder) --[chan]--> (Rest of the logs pipeline...)
+* (devices) --[udp:portN]-!-> (Listener N) /
+*
+* ! = delimitation between the outside world and the Agent
+ */
+
+// SnmpPacket is the type of packets yielded by server listeners.
+type SnmpPacket struct {
+	Content *gosnmp.SnmpPacket
+	Addr    *net.UDPAddr
+}
+
+// OutputChannel is the type of the server output channel.
+type OutputChannel = chan *SnmpPacket
+
+// TrapServer runs multiple SNMP traps listeners.
+type TrapServer struct {
+	Started bool
+
+	bindHost    string
+	listeners   []TrapListener
+	output      OutputChannel
+	stopTimeout time.Duration
+}
+
+var (
+	outputChannelSize = 100
+	// RunningServer holds a reference to the trap server instance running in the Agent.
+	RunningServer *TrapServer
+)
+
+// NewTrapServer configures and returns a running SNMP traps server.
+func NewTrapServer() (*TrapServer, error) {
+	var configs []TrapListenerConfig
+	err := config.Datadog.UnmarshalKey("snmp_traps_listeners", &configs)
+	if err != nil {
+		return nil, err
+	}
+
+	bindHost := config.Datadog.GetString("bind_host")
+	stopTimeout := config.Datadog.GetDuration("snmp_traps_stop_timeout") * time.Second
+
+	output := make(OutputChannel, outputChannelSize)
+	listeners := make([]TrapListener, 0, len(configs))
+
+	for _, c := range configs {
+		listener, err := NewTrapListener(bindHost, c, output)
+		if err != nil {
+			return nil, err
+		}
+
+		listeners = append(listeners, *listener)
+	}
+
+	s := &TrapServer{
+		Started:     false,
+		bindHost:    bindHost,
+		listeners:   listeners,
+		output:      output,
+		stopTimeout: stopTimeout,
+	}
+
+	err = s.start()
```

Comment:
Here if we have a single listener that fails, all listener, including valid ones, will be ignored.

Reply:
I think this is now fixed! Invalid listeners are just ignored, allowing valid ones to start.